### PR TITLE
feat(cdk/testing): support waiting for tasks outside angular in protractor harness environment

### LINKED ISSUE #18375: Add support for `waitForTasksOutsideAngular` in protractor tests
Currently this has only been implemented in `TestbedHarnessEnvironment`, it should be implemented for `ProtractorHarnessEnvironment` too. I think we can reuse the same infrastructure we created to deal with this in unit tests, but just run it as a browser script in protractor. See https://github.com/angular/components/pull/18361 for some additional discussion

### PR DESCRIPTION
**Pull request by devversion**
We added support to harnesses for waiting until all tasks outside Angular
complete. We only added this to the `TestBed` harness environment because
it was unclear how we can get this functionality work in e2e tests.

Investigation showed that we can get it to work in a similar matter
by leveraging ZoneJS. Ideally we would only inject necessary code to
intercept ZoneJS dynamically through `browser.executeScript`, but that
does not work because we'd need to intercept the root zone for e2e
tests, unlike the proxy zone in TestBed jasmine tests. The root zone
does not keep track of the current task state, so if we intercept
at any time, we do not know if there are previously scheduled tasks
that will throw off our interceptor.

Due to this, we need to ask harness consumers in e2e tests/Protractor
tests to load a separate file before loading ZoneJS. This file will
intercept ZoneJS immediately so that the e2e/protractor harness
environment can then hook into the intercepted root zone.

More details can be found in the dedicated readme attached to this
commit. See `src/cdk/testing/zone-task-interceptor/README.md`.

Resolves COMP-333. Closes #18375.

### CODE REVIEW

**Comment by devversion** [comment]
The question for this PR is: Do we want to require e2e harness consumers to load a file together with ZoneJS in order to be able to wait for tasks outside of the Angular zone. I dived deep into ZoneJS and there  is _no way_ to get it work without loading a file initially together with ZoneJS (as proposed with this PR). Alternatively we could potentially land a change in ZoneJS that would allow us to do that. Though I have a feeling this is unlikely to happen any time soon, and also it would affect ZoneJS in production applications.

I'd think loading a separate file is totally fine, given that for unit tests, one would need to load the ZoneJS testing bundle too. At the same time though, I'd expect the e2e harnesses to work in production applications (without needing to load a separate file).

File: src/cdk/testing/tests/test-main-component.ts [hunk 1]
```diff
@@ -104,7 +104,9 @@ export class TestMainComponent implements OnDestroy {
   }
 
   runTaskOutsideZone() {
+    console.error('Start');
     this._zone.runOutsideAngular(() => setTimeout(() => {
+      console.error('Timeout ');
```
Comment (devversion) [inline src/cdk/testing/tests/test-main-component.ts:109]: Need to remove these debugging statements

**Comment by LayZeeDK** [comment]
> The question for this PR is: Do we want to require e2e harness consumers to load a file together with ZoneJS in order to be able to wait for tasks outside of the Angular zone? I dived deep into ZoneJS and there is _no way_ to get it work without loading a file initially together with ZoneJS (as proposed with this PR). Alternatively we could potentially land a change in ZoneJS that would allow us to do that. Though I have a feeling this is unlikely to happen any time soon, and also it would affect ZoneJS in production applications.

Did you consult @JiaLiPassion?

> I'd think loading a separate file is totally fine, given that for unit tests, one would need to load the ZoneJS testing bundle too. At the same time though, I'd expect the e2e harnesses to work in production applications (without needing to load a separate file).

I went through an Angular workspace freshly generated by the Angular CLI. For Karma unit tests we have `test.ts`, but I don't see extra files being loaded as part of Protractor tests. What would that look like? An `e2e` configuration for the `serve` command which has a `fileReplacements` option for `polyfills.ts` to replace it with a `polyfills.e2e.ts` file as part of that build?

**Comment by mmalerba** [comment]
@LayZeeDK we're hoping not to have to do this. See https://github.com/angular/angular/issues/32896 for discussion of potential changes in rxjs that would allow us to do this more simply.

**Comment by devversion** [comment]
@LayZeeDK Thanks for the initial feedback already! We are currently working on a solution that _might_ be suitable to land in ZoneJS itself. See: https://github.com/angular/angular/issues/32896 for discussion if you are interested, and an [experiment patch](https://github.com/devversion/angular/commit/71d0260d12f79d19f71028ebbf62a91c2a6b4d3f) that would already do the trick for our concrete use-case at least.

Also, great example with the CLI configuration. I think the changes you've made there are already reasonable, but definitely a bit cumbersome and inconvenient to maintain/implement (that could be handled by an schematic if we _need_ to go with that route).

The problem we have right now is that there seems really no way to retrieve the task state from the root zone before we intercept the root zone. This means that we would need a patch to be loaded together with ZoneJS on init, or we need ZoneJS to keep the task count in memory by default. That would solve the problems nicely but unfortunately _might_ not be do-able as it requires a change in ZoneJS purely for our use-case. I'm hoping we can work out an API that might be useful and general and is reasonable enough to build into ZoneJS by default (with low perf and memory impact; i.e. only keeping three numbers in memory for the root zone).

Also we should acknowledge that this won't work nicely for Angular applications using the Angular noop NgZone. IIRC though, Protractor w/ the blocking-proxy does not work with the Angular noop zone either.

**Comment by devversion** [comment]
This is obsolete with the protractor harness environment being deprecated.

**Pull request closed without merging**